### ISSUE TITLE
[0006] Make `SPV_KHR_variable_pointers` mandatory for the SPIR-V output

### ISSUE THREAD

**Issue by devshgraphicsprogramming** [microsoft/hlsl-specs#99]
**Which proposal does this relate to?**
https://github.com/microsoft/hlsl-specs/blob/main/proposals/0006-reference-types.md

**Describe the issue or outstanding question.**

clspv implements pointers thanks to the use of `SPV_KHR_variable_pointers`

Even without pointers you can make an abomination that requires you to have the same problem
```cpp
int& choose(int& a, int& b, bool which) {
    return which ? a:b;
}
```

**Additional context**

Here's a quick explainer on why this was a mandatory feature for CLSPV https://github.com/google/clspv/issues/113

Btw 98% of all Vulkan 1.2 devices have it, the only notable recently reported exceptions are the "software" (LLVMpipe/SwiftShader) and "emulated" (Vk on D3D12) devices.

Not sure whats the solution for DXIL here.

**Comment by devshgraphicsprogramming** [comment]
I came back having learnt that `SPV_KHR_variable_pointers` doesn't provide for `Private` storage class variables, so anything that can "leak" would need to be demoted to some dynamically indexed and type punned array equivalent to
```cpp
union
{
   uint32_t _uint32_t[LEAKED_PRIVATE_MEM_SIZE/sizeof(uint32_t)];
   float32_t _float[LEAKED_PRIVATE_MEM_SIZE/sizeof(float32_t)];
   ... all the other types you use
};
```
But I can also see in another issue #49 it has been pointed out that there's no way to get this done current SPIR-V.

If you are operating under the limitation of disallowing type punning, then I guess you'd just have these arrays be non-unionized and with sepearate sizes, so this turns into a DOS-era strict-C no-unified-address-space as if you had an `alloca<T>` which is sensitive to the type its allocating for and each time has its own address space.

Obviously yet another compromisse/backing-down is possible which is to disallow leaks, so no choosing/ternary operators, storing references into const arrays which are non-constexpr indexed, etc.

However if you allow leaking of function references you'll have to do it the non-type-punned way + an ugly unrolled search-binary-tree dispatcher (cause you can't count on `switch` being lowered to a binary search in for example NIR), unless something in SPIR-V changes:
https://github.com/KhronosGroup/Vulkan-Docs/issues/2232

TL;DR SPIR-V has a lot of issues and I'd view it as an opportunity to steer DXIL clear of those during its evolution.

I mean this kinda makes sense, cause `Private` memory should be by default, thought of to live in "registers" (no way to get the address, index, also variables may move between regs thanks to SSA, spilling, etc.) and have quite short/trimmed lifetimes.

**Comment by devshgraphicsprogramming** [comment]
I'm not a SPIR-V expert but it seems like `OpSelect` can pick between two Logical Pointers, does current DXIL have any way to express this?
```
    // syntax for non-dynamic statically knowable references/pointers
    int varA = 0;
    int varB = 2;
    
    // in C++ the address that a reference points to is immutable (cannot change)
    auto& refVar = (argc%2) ? varA:varB; // but there's nothing stopping me from initializing it in a not-compile time knowable way
    
    // one would think this would work, but addresses of `varA` and `varB` are not known at compile time, due to alloca (maybe in HLSL they could be knowable because they'd be logical, but only at link time)
    // error: constexpr variable 'compileRefVar' must be initialized by a constant expression
    //constexpr auto& compileRefVar = varA;

    // we can only grab references to global variables :(
    // constexpr auto& compileRefVar = argc;
    constexpr auto& compileRefVar = globalVarA;
    compileRefVar = 5;     
    std::cout << compileRefVar << "," << globalVarA << std::endl;

    // thankfully Clang refuses this dynamic/non-knowable reference assignment
    // constexpr auto& compileRefVar = (argc%2) ? globalVarA:globalVarB;
    // error: constexpr variable 'compileRefVar' must be initialized by a constant expression

    // it will also refuse for a pointer, with nice diagnostics
    // error: constexpr variable 'compilePtr' must be initialized by a constant expression
    // note: function parameter 'argc' with unknown value cannot be used in a constant expression
    //constexpr auto* compilePtr = (argc%2) ? &globalVarA:&globalVarB;

    // however same issue persists as with references
    // error: constexpr variable 'compilePtr' must be initialized by a constant expression
    //constexpr auto* compilePtr = &varA;

    // but it works just as well as a reference, its a bit weird that the syntax is not `auto* constexpr` for a constexpr pointer to a mutable
    constexpr auto* compilePtr = &globalVarA;
    *compilePtr = 5;
    std::cout << compileRefVar << "," << globalVarA << std::endl;
    
    // good just as we want it
    // error: cannot assign to variable 'compilePtr' with const-qualified type 'int *const'
    //compilePtr = nullptr;
```
and even if you extend DXIL, will this always work for a variable of any type? even opaque non-storable ones?

I know it definitely won't work for function references in SPIR-V.


I don't know if `constexpr T&` will cut it, as function parameters just can't be `constexpr`
```cpp
int functionThatTakesAConstexprRef(constexpr int& num)
{
    return num * num;
}
```
so you end up needing to use a silly syntax like so
```cpp
template<int& num>
int functionThatTakesAConstexprRef()
{
    return num * num;
}
```

This will all force you to need to derive from _at least_ C++17.

Plus for the whole thing to work you'd need to "pretend" that local variables in functions do indeed have addresses known at compile time, otherwise this only works for global and static variables and function addresses.
